pkg/gitprovider/bitbucket.go

```diff
@@ -62,49 +62,59 @@ func (g *BitbucketGitProvider) GetRepositories(namespace string) ([]*GitReposito
 		namespace = user.Username
 	}
 
-	repoList, err := client.Repositories.ListForAccount(&bitbucket.RepositoriesOptions{
-		Owner:   namespace,
-		Page:    &[]int{1}[0],
-		Keyword: nil,
-	})
-	if err != nil {
-		return nil, err
-	}
-
-	for _, repo := range repoList.Items {
-		htmlLink, ok := repo.Links["html"].(map[string]interface{})
-		if !ok {
-			return nil, fmt.Errorf("Invalid repo links")
-		}
-
-		repoUrl, ok := htmlLink["href"].(string)
-		if !ok {
-			return nil, fmt.Errorf("Invalid repo html link")
-		}
-
-		u, err := url.Parse(repoUrl)
+	// Since bitbucket doesnt supports perPage arg,
```

Comment:
You can construct your own pagination here.

If the default is always 10, then you should fetch until `perPage` is satisfied.

Now you fetch the entire list every time which is not ideal.

Reply:
> Now you fetch the entire list every time which is not ideal.
I am fetching the list based on **Page** option.  

Cant rely on **perPage** here since then how to keep track of the page no. ? Say, we exhausted first iteration (from frontend **For** loop call with 1st page 100 items), on next call it will be 2nd page with next 100 items, but on backend we are not aware which page no. we should start from (or we can say 10*10 items so start from 11th page? Not sure if we should go with this approach ?)

Rather I just exhaust the items in one iteration and break the next frontend iteration.

```
repoList, err := client.Repositories.ListForAccount(&bitbucket.RepositoriesOptions{
			Owner: namespace,
			Page:  &options.Page,
		})
...
options.Page++
```

Reply:
Had a doubt now if its really 10 items as default, found out in bitbucket's pkg code [here](https://github.com/ktrysmt/go-bitbucket/blob/master/client.go#L23). Seems we can change the default size, but lets leave that ...

Reply:
![Screenshot 2024-07-26 at 09 48 37](https://github.com/user-attachments/assets/35aef29d-ace2-4201-9eea-fbe0f1b69513)

Good that you linked their code. Nice digging 😄 
You can then just do `client.Pagelen` and that's it!

Reply:
Should we do that ? client.Pagelen = 100 (perPage can change in future and someone misses this, thou highly unlikely..)

Reply:
on second thoughts, lets do it!

Reply:
@abhishek818 you should do it per-request, not on a "global" level. In each method you get the client, you can then set the pagelen based on the parameter of the method and that's it.

Reply:
fixed!

Reply:
wait!

Reply:
> @abhishek818 you should do it per-request, not on a "global" level. In each method you get the client, you can then set the pagelen based on the parameter of the method and that's it.

WTH obviously that should be the way, i am messing up big time..

Reply:
need your suggestion , only fetch api for bitbucket has Page arg, other like fetch namespaces, PRs, branches dont have a Page arg, should i just set Pagelen to some big integer like 999 ?

Reply:
```golang
func (g *BitbucketGitProvider) GetRepoPRs(repositoryId string, namespaceId string,  options ListOptions) ([]*GitPullRequest, error) {
	client := g.getApiClient()
```

This would become
```golang
func (g *BitbucketGitProvider) GetRepoPRs(repositoryId string, namespaceId string,  options ListOptions) ([]*GitPullRequest, error) {
	client := g.getApiClient()
        client.Pagelen = options.PerPage
```

Reply:
I get that, what i meant was this should be done only for the getrepos function, other apis dont have a Page arg also (unlike getrepos). 
So I feel for other except fetchrepos, we should set pagelen to something like 1000 ?

Reply:
```
func (g *BitbucketGitProvider) GetRepositories(namespace string, options ListOptions) ([]*GitRepository, error) {
	client := g.getApiClient()
	// Bitbucket doesnt supports perPage arg in fetch apis request
	client.Pagelen = options.PerPage
```

Others:
```
func (g *BitbucketGitProvider) GetNamespaces(options ListOptions) ([]*GitNamespace, error) {
	client := g.getApiClient()
	client.Pagelen = 1000
```

Reply:
had pushed this change, its alright , feel free to comment when you get time..

Reply:
Ah, okay. You can just ignore the param in that case since it doesn't make sense with a Page arg.

Reply:
you mean not using Pagelen for them or setting that to 1000 ?

Reply:
alright then, up for review.

Reply:
Not using Pagelen since it's irrelevant I believe.

Please check in the bitbucket lib. If it is relevant, you can set 1000 but then set it in the `getApiClient` method with a comment on why that is.

Reply:
done, pagelen is relevant for apis. Have added the code comment, refer [link](https://github.com/ktrysmt/go-bitbucket/blob/master/client.go#L259)

Reply:
up for review!